### ISSUE TITLE
[QUESTION] the structure of gpt is "encoder-only" in megatron

### ISSUE THREAD

**Issue by 1926627357** [NVIDIA/Megatron-LM#439]
Megatron-LM provides code file of gpt in [gpt_model.py](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/gpt_model.py#L43). It builds gpt model by `get_language_model` [function](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/language_model.py#L51) while setting `add_encoder=True` and `add_decoder=False` as default. Actually, gpt model is decoder-only on the contrary. Therefore, I'm very confused now whether gpt model provided by megatron is correct, or some infomation is overlooked by myself.  :(

**Comment by janEbert** [comment]
Similar to the PyTorch `torch.nn.TransformerDecoder` module, here, "decoder" implies usage of cross-attention between the encoder outputs and decoder intermediary outputs. Of course, this only matters if you have both an encoder and a decoder (i.e., you are using an encoder-decoder Transformer). Now going away from encoder-decoder Transformers, whether you're considering an encoder-only or a decoder-only architecture, they are actually completely the same; only the attention mask is different: full for encoder-only, causal for decoder-only.

So whether you use Megatron's `ParallelTransformerLayer([...], add_encoder=True, add_decoder=False)` or PyTorch's `torch.nn.TransformerEncoder`, these are the only ways you can obtain a decoder-only model. Again, this is due to the expectations that the "decoder" version of the layer/module has: it expects to receive an encoder's output to perform cross-attention on.

**Comment by haiqwa** [comment]
I see! Thanks!

**Comment by bkkm78** [comment]
I believe this is a very common confusion about the naming of these models. I really think people should come up with more descriptive names. Nevertheless, here is a summary of what the terms really mean in different context:

Original transformer paper (_Attention is all you need_):
- Encoder: no cross attention, no causal masks.
- Decoder: with cross attention, with causal masks.

"Encoder-only" vs "decoder-only" models:
- "Encoder": no causal masks.
- "Decoder": with causal masks.
- "-Only": no cross attention.

PyTorch `TransformerEncoderLayer`, `TransformerDecoderLayer`:
- Encoder: no cross attention, optional causal masks.
- Decoder: with cross attention, optional causal masks.

So to create a GPT-like "decoder-only" model using PyTorch's modules or libraries with similar naming conventions, one needs to use the encoder modules instead of the decoder modules, and set up the causal masks to make it "decoder-only".
